Optional file behavior

neurolabusc opened the pull request:
This is a very simple patch that allows developers to modify the behavior of this library. All are disabled by default, so no change for tools unless they invoke these features.

This provides three compile time options to influence how file handling behavior

1. Related to [issue 129](https://github.com/NIFTI-Imaging/nifti_clib/issues/129). Emulate FSL file handing behavior. Refuse to open `T1.nii` if file `T1.nii.gz` exists. Likewise, refuse to open `T1.nii.gz` if file `T1.nii` exists. This feature  enabled with the directive:
```
#define FSLSTYLE 
```
A conflict generates an error:
```
Multiple possible filenames detected for basename (*.nii, *.nii.gz): T1.nii.gz
```
2. Use pigz parallel compression when writing .nii.gz files if the environment includes the variable `AFNI_COMPRESSOR=PIGZ` and the pigz executable is found. This behavior is enabled by the compiler directive:

```
#define PIGZ
```

3. Refuse to open files of complex datatype. Many tools are not designed for handling complex data. This option is enabled by the compiler directive:

```
#define REJECT_COMPLEX
```

neurolabusc committed:
emulate FSL (.nii/.nii.gz), emulate AFNI pigz, reject complex data (https://github.com/NIFTI-Imaging/nifti_clib/issues/129)

neurolabusc committed:
Disable defaults

gdevenyi commented:
Only request I would have is that the first commit be split into relevant separate changes rather than a single commit.

gdevenyi commented:
Oh and it looks like there's no modification to the configure or cmake to expose these options.

neurolabusc committed:
Configure 'run "FSL=1 make"' and 'cmake -DFSLSTYLE=ON .'  (https://github.com/NIFTI-Imaging/nifti_clib/pull/138)

neurolabusc commented:
@gdevenyi I don't think this needs to be three commits - the code uses `ifdef` blocks to isolate each. I think the vast majority of developers will want to switch them on/off together. Only advanced cases would separate them (e.g. a tool that mimics `fslcomplex` would want to allow complex data types; while most scalar processing tools will want to reject these images). The `PIGZ` option is dormant unless the user explicitly sets an environment variable.  

I have added options to allow the user to switch on these features:

To compile with cmake, use the `FSLSTYLE` option:
```
cmake -DFSLSTYLE=ON ..
```
To compile with make in the `nifti2` folder:

```
run "FSL=1 make"
```

To demonstrate, assume the folder contains two files `a.nii` and `a.nii.gz`. Both fslmaths and nifti_tool will reject this:

```
nifti2 > fslmaths ~/a -add 13 anew
Image Exception : #61 :: Multiple possible filenames detected for basename: a
libc++abi: terminating with uncaught exception of type std::runtime_error: Multiple possible filenames detected for basename: a
zsh: abort      fslmaths a -add 13 anew

nifti2 > ./nifti_tool -mod_hdr -prefix anew -infiles a -mod_field descrip 'beer, brats and cheese, mmmmm...'

Image Exception : Multiple possible filenames detected for basename (*.nii, *.nii.gz): a.nii.gz
```

I added these features to my open source fslmaths clone [niimath](https://github.com/rordenlab/niimath). Selfishly, having this patch included with nifti_clib will allow my to ease keeping my project up to date. However, given the popularity of FSL tools, I do think these features will insulate users from confusion.

neurolabusc committed:
Make names consistent

neurolabusc committed:
nifti_strdup() is too small to include an extension

neurolabusc commented on nifti2/nifti2_io.c:
```diff
@@ -3741,7 +3746,23 @@ char * nifti_findhdrname(const char* fname)
 
    strcpy(hdrname,basename);
    strcat(hdrname,elist[efirst]);
+   #ifdef FSLSTYLE
```
I tried to mimic the FSL behavior, where having a naming conflict (e.g. 'filename.nii' and 'filename.hdr') generates a hard error. I like your idea of generating a warning if this is not the case. Perhaps you want to add this in a way that will be familiar to AFNI users.

afni-rickr reviewed (commented):
How would you prefer to proceed?  Would you like me to add a function to detect name conflicts?  Would you like to?  Thanks.

neurolabusc commented on nifti2/nifti2_io.c:
```diff
@@ -5938,6 +5959,13 @@ nifti_image *nifti_image_read( const char *hname , int read_data )
       znzclose(fp);  free(hfile);  return NULL;
    }
 
+   #ifdef REJECT_COMPLEX
```
Feel free to adapt as you wish. My goal was to emulate fslmaths behavior, and the immediate exit() allows us to mimic the same error messages. If a developer wants to reject complex data, I do think that exiting here is acceptable, as this is a terminal error for these use cases.

neurolabusc commented on nifti2/nifti2_io.c:
```diff
@@ -7770,6 +7806,73 @@ znzFile nifti_image_write_hdr_img( nifti_image *nim , int write_data ,
  do{ fprintf(stderr,"** ERROR: nifti_image_write_hdr_img: %s\n",(msg)) ;  \
      return fp ; } while(0)
 
+#ifdef PIGZ
```
My goal here was to allow the developer to emulate AFNI behavior, using the same environment variable a user could specify for AFNI. Feel free to rename or modify to follow the conventions of this library.

afni-rickr commented:
@neurolabusc Sure, this seems good.  To proceed, so that your commits are recorded and I can add to them, would it be okay for me to make a new branch and pull your changes into it?  Then I can make changes in the branch before seeing whether we are happy to merge into master.  Thanks!

neurolabusc commented:
Sounds good.

hjmjohnson commented:
@afni-rickr Can the requests from this PR be addressed soon?

seanm commented:
@neurolabusc this was merged over in the ITK fork of nifti_clib: https://github.com/InsightSoftwareConsortium/nifti_clib/pull/9

I did have a few questions in that PR that I wonder if you might be able to answer...?

neurolabusc commented:
@seanm feel free to ask questions or email for a zoom meet.

@afni-rickr I wonder if we should just use the ITK fork as a PR to this main repo

seanm commented:
> @seanm feel free to ask questions or email for a zoom meet.

@neurolabusc my questions are here: https://github.com/InsightSoftwareConsortium/nifti_clib/pull/9#pullrequestreview-2581947731  Would be great if you could reply there.  I can make a followup PR to clean things up a bit based on your responses.
